### PR TITLE
Synth-check and accelerate options for sygus-rr

### PR DESCRIPTION
**Pull request by ajreynol**
Adds two new options to the sygus-rr feature:
- The option sygus-rr-synth-accel adds dynamic symmetry breaking lemma templates so that the sygus search space is pruned based on the assumption that the term on one side of candidate-rewrite is redundant. 
-  The option sygus-rr-synth-check uses a separate satisfiability query to check whether candidate-rewrites are true rewrites.  With this option, candidate-rewrites are either discarded, or verified correct via the query t1 != t2, where if this is UNSAT we print:
(rewrite t1 t2)
instead of 
(candidate-rewrite t1 t2)
This uses model-based sampling to reduce the rate of false positives, e.g. counterexample points to candidate-rewrites are incrementally added to the sygus sampler.

### CODE REVIEW

File: src/theory/quantifiers/sygus/ce_guided_conjecture.cpp [hunk 1]
```diff
@@ -615,28 +617,97 @@ void CegConjecture::printSynthSolution( std::ostream& out, bool singleInvocation
...
-              Node eq_solb = sygusDb->sygusToBuiltin(eq_sol);
-              Node eq_solr = er->extendedRewrite(eq_solb);
-              Trace("sygus-rr-debug")
-                  << "; candidate #1 ext-rewrites to: " << solbr << std::endl;
-              Trace("sygus-rr-debug")
-                  << "; candidate #2 ext-rewrites to: " << eq_solr << std::endl;
+              // The analog of terms sol and eq_sol are equivalent under sample
+              // points but do not rewrite to the same term. Hence, this
+              // indicates
+              // a candidate rewrite.
```
Comment (HanielB) [inline src/theory/quantifiers/sygus/ce_guided_conjecture.cpp]: why this break line?

File: src/theory/quantifiers/sygus/ce_guided_conjecture.cpp [hunk 2]
```diff
@@ -615,28 +617,97 @@ void CegConjecture::printSynthSolution( std::ostream& out, bool singleInvocation
...
+            Node eq_solb = sygusDb->sygusToBuiltin(eq_sol);
+            Node eq_solr = er->extendedRewrite(eq_solb);
+            bool success = true;
+            bool verified = false;
+            // verify it if applicable
+            if (options::sygusRewSynthCheck())
+            {
+              Trace("rr-check") << "Check candidate rewrite..." << std::endl;
+              SmtEngine rrChecker(NodeManager::currentNM()->toExprManager());
+              // rrChecker.setOption("produce-models", SExpr("true"));
```
Comment (HanielB) [inline src/theory/quantifiers/sygus/ce_guided_conjecture.cpp]: kill this line

File: src/theory/quantifiers/sygus/ce_guided_conjecture.cpp [hunk 3]
```diff
@@ -615,28 +617,97 @@ void CegConjecture::printSynthSolution( std::ostream& out, bool singleInvocation
...
+                d_sampler[prog].getVariables(vars);
+                std::vector<Node> pt;
+                for (const Node& v : vars)
+                {
+                  Node val = Node::fromExpr(rrChecker.getValue(v.toExpr()));
+                  Trace("rr-check") << "  " << v << " -> " << val << std::endl;
+                  pt.push_back(val);
+                }
+                d_sampler[prog].addSamplePoint(pt);
+                // readd the solution
```
Comment (HanielB) [inline src/theory/quantifiers/sygus/ce_guided_conjecture.cpp]: readd --> read

Comment (ajreynol) [reply src/theory/quantifiers/sygus/ce_guided_conjecture.cpp]: I do mean "re-add" :)

Comment (ajreynol) [reply src/theory/quantifiers/sygus/ce_guided_conjecture.cpp]: Changed to "add this solution again"

**Pull request merged**